### PR TITLE
[test] Add urlencoded form

### PR DESCRIPTION
**Pull request by chamons**
- Part of https://github.com/azure/autorest.csharp/issues/986
- .NET test that will land later - https://gist.github.com/chamons/76c377098cb30e2dc8b6214ebf3326f2

### CODE REVIEW

File: swagger/body-formdata-urlencoded.json [hunk 1]
```diff
@@ -53,5 +59,33 @@
...
+        "description": "Can take a value of dog, or cat, or fish",
+        "type": "string",
+        "in": "formData",
+        "required": true,
+        "enum": [
+            "dog",
+            "cat",
+            "fish"
+        ],
+        "x-ms-parameter-location": "method"
```
Comment (pakrym) [inline swagger/body-formdata-urlencoded.json:75]: let's make one of these enums extensible to get a bit more coverage.

Comment (pakrym) [reply swagger/body-formdata-urlencoded.json:75]: see https://github.com/stankovski/AutoRest/blob/master/Documentation/swagger-extensions.md#x-ms-enum

**Comment by pakrym** [comment]
Rev https://github.com/Azure/autorest.testserver/blob/master/package.json#L3

**Review by pakrym — APPROVED** [review]
Sweet!

File: src/api/request-validations.ts [hunk 2]
```diff
@@ -85,3 +85,10 @@ export const validateXMLBodyEquals = async (request: RequestExt, expectedBody: s
     throw new ValidationError(BODY_NOT_EQUAL_ERROR_MESSAGE, expectedParsedBody, actualParsedBody);
   }
 };
+
+export const validateHeader = (request: RequestExt, header_name: string, expected: string) => {
+  const actual = request.headers[header_name];
+  if (actual != expected) {
```
Comment (timotheeguerin) [inline src/api/request-validations.ts:91]: realize there is no linting rule for that, so I'll one but you should pretty much always use triple `=` in javascript
`===` or `!==` . Using `==` has unexpected consequences

File: swagger/body-formdata-urlencoded.json [hunk 3]
```diff
@@ -28,6 +28,12 @@
             "required": true,
             "type": "string"
```
Comment (pakrym) [inline swagger/body-formdata-urlencoded.json:29]: let's make this an int, we already have string parameters.

File: src/api/request-validations.ts [hunk 4]
```diff
@@ -85,3 +85,10 @@ export const validateXMLBodyEquals = async (request: RequestExt, expectedBody: s
...
+    throw new ValidationError(`Expected ${expected} but got ${actual}`, undefined, request.headers);
```
Comment (timotheeguerin) [inline src/api/request-validations.ts:92]:
```suggestion
    throw new ValidationError(`Expected ${expected} but got ${actual}`, expected, actual);
```

2nd argument is the expected value so I would do that

File: src/test-routes/formdata.ts [hunk 5]
```diff
@@ -14,4 +15,11 @@ app.category("vanilla", () => {
       },
     };
   });
+  app.post("/formsdataurlencoded/pet/add", "UpdatePetWithForm", (request) => {
+    validateHeader(request.originalRequest, "content-type", "application/x-www-form-urlencoded");
```
Comment (timotheeguerin) [inline src/test-routes/formdata.ts:19]: I would add this to the `request-expectation.ts` so instead you can use those method `request.expect.validateHeader`

for `validateBodyEquals` below you should already be able to.

Comment (pakrym) [reply src/test-routes/formdata.ts:19]: Should all methods from `request-validations` go there (not as part of this or)?

Comment (chamons) [reply src/test-routes/formdata.ts:19]: It appears that RequestExpectation is just scoped trampolines calling those methods. I'll just add one for the header.

**Review by timotheeguerin — APPROVED** [review]

**Comment by chamons** [comment]
Waiting for @pakrym second 👍

File: src/test-routes/formdata.ts [hunk 6]
```diff
@@ -14,4 +15,11 @@ app.category("vanilla", () => {
       },
     };
   });
+  app.post("/formsdataurlencoded/pet/add", "UpdatePetWithForm", (request) => {
+    request.expect.containsHeader("content-type", "application/x-www-form-urlencoded");
+    request.expect.bodyEquals({ pet_type: "dog", pet_food: "meat", name: "Fido", status: "" });
```
Comment (pakrym) [inline src/test-routes/formdata.ts:20]: let's validate pet_id as well

**Review by pakrym — APPROVED** [review]

File: src/test-routes/formdata.ts (hunk 6 above)
Comment (chamons) [reply src/test-routes/formdata.ts:20]: Good catch!

I wasn't correctly generating code to pass it!

**Pull request merged**